includes/class-wcs-action-scheduler-customer-notifications.php

```diff
@@ -0,0 +1,290 @@
+			return;
+		}
+
+		as_schedule_single_action( $timestamp, $action, $action_args );
+	}
+
+	//TODO: check timezones
+	/*
+	 * Subtract time offset from given datetime based on the settings and subscription properties.
+	 *
+	 * @param string $datetime
+	 * @param WC_Subscription $subscription
+	 *
+	 * @return int
+	 */
+	protected function sub_time_offset( $datetime, $subscription ) {
+		$dt = new DateTime( $datetime, new DateTimeZone( 'UTC' ) );
+		$dt->sub( new DateInterval( "PT{$this->get_hours_offset( $subscription )}H" ) );
+
+		return $dt->getTimestamp();
+	}
+
+	public function schedule_trial_ending_notification( $subscription ) {
+		$trial_end = $subscription->get_date( 'trial_end' );
+		$timestamp = $this->sub_time_offset( $trial_end, $subscription );
+
+		$this->schedule_notification(
+			$subscription,
+			'woocommerce_scheduled_subscription_customer_notification_trial_expiration',
+			$timestamp
+		);
+	}
+
+	public function schedule_expiry_notification( $subscription ) {
+		$subscription_end = $subscription->get_date( 'end' );
+		$timestamp        = $this->sub_time_offset( $subscription_end, $subscription );
+
+		$this->schedule_notification(
+			$subscription,
+			'woocommerce_scheduled_subscription_customer_notification_expiration',
+			$timestamp
+		);
+	}
+
+	public function schedule_payment_notification( $subscription ) {
+
+		//TODO: For end of trial, should we schedule payment notification? Let's say no.
+		if ( $subscription->get_date( 'trial_end' ) ) {
+			return;
+		}
+
+		$next_payment = $subscription->get_date( 'next_payment' );
+		$timestamp    = $this->sub_time_offset( $next_payment, $subscription );
+
+		// Can manual vs automatic payment change until it runs?
+		if ( $subscription->is_manual() ) {
+			$this->schedule_notification(
+				$subscription,
+				'woocommerce_scheduled_subscription_customer_notification_manual_renewal',
+				$timestamp
+			);
+		} else {
+			$this->schedule_notification(
+				$subscription,
+				'woocommerce_scheduled_subscription_customer_notification_auto_renewal',
+				$timestamp
+			);
+		}
+	}
+
+	/**
+	 * Maybe set a schedule action if the new date is in the future
+	 *
+	 * @param object $subscription An instance of a WC_Subscription object
+	 * @param string $date_type Can be 'trial_end', 'next_payment', 'payment_retry', 'end', 'end_of_prepaid_term' or a custom date type
+	 * @param string $datetime A MySQL formatted date/time string in the GMT/UTC timezone.
+	 */
+	public function update_date( $subscription, $date_type, $datetime ) {
+
+		if ( in_array( $date_type, $this->get_date_types_to_schedule(), true ) ) {
+
+			if ( $subscription->get_date( 'trial_end' ) ) {
+				$this->schedule_trial_ending_notification( $subscription );
+			}
+
+			if ( $subscription->get_date( 'end' ) ) {
+				$this->schedule_expiry_notification( $subscription );
+			}
+
+			if ( $subscription->get_date( 'next_payment' ) ) {
+				$this->schedule_payment_notification( $subscription );
+			}
```

Comment:
This could be right but wanted to double check. When a subscription's date is updated you reschedule all dates? 

eg if the next payment date is updated, it would pass the `in_array()` check and then all subscription dates would be rescheduled.

Reply:
> When a subscription's date is updated you reschedule all dates?
Yes. If the scheduling is correct, nothing is written to the db, but the code currently checks whether the scheduling is correct for all 3 types of notifications (and fixes it if it isn't).